involute_gears.scad

```diff
@@ -42,488 +42,538 @@ pi=3.1415926535897932384626433832795;
+
+    //Outer Circle
+    outer_radius = pitch_radius+addendum;
+
+    // Dedendum: Radial distance from pitch circle to root diameter
+    dedendum = addendum + clearance;
+
+    // Root diameter: Diameter of bottom of tooth spaces.
+    root_radius = pitch_radius-dedendum;
+    backlash_angle = backlash / pitch_radius * 180 / pi;
+    half_thick_angle = (360 / number_of_teeth - backlash_angle) / 4;
+
+    // Variables controlling the rim.
+    rim_thickness = (rim_thickness!=undef?(rim_thickness!=0?rim_thickness:gear_thickness):gear_thickness * 1.5);
+    rim_width = (rim_width!=undef?rim_width:root_radius * .1);
+    rim_radius = root_radius - rim_width;
+
+    // Variables controlling the hub_diameter
+    hub_thickness = (hub_thickness!=undef?(hub_thickness!=0?hub_thickness:gear_thickness):gear_thickness * 2);
+    hub_diameter = (hub_diameter!=undef?hub_diameter:root_radius * .3);
+
+    // Variables controlling the spokes
+    spokes = spokes == undef? 0 : spokes;
+    spoke_thickness = (spoke_thickness == undef)? rim_thickness : spoke_thickness;
+    spoke_width = (spokes==0)? 1 : (spoke_width == undef)?  0.75 * pi * hub_diameter / spokes : spoke_width; 
+    //spoke_depth is depth spoke must penetrate into hub to ensure complete penetration
+    spoke_depth = ((hub_diameter/2)^2-(spoke_width/2)^2)^0.5 +.01;
+    //spoke length is length of spoke including the depth sunk into the hub
+    spoke_length = spoke_depth+rim_radius-(hub_diameter/2.0);
+    //spoke raius is the distance from gear center to base of the spoke(inside the hub)
+    spoke_radius = (hub_diameter/2.0)-spoke_depth;
+    //echo (str("spoke_width: ",spoke_width,", hub_diameter: ",hub_diameter, ", spoke_depth: ",spoke_depth));
+    
+    // Variables controlling the bore
+    bore_diameter = (bore_diameter!=undef?bore_diameter:root_radius * .1);
+
+    // Variables controlling the circular holes in the gear.
+    circles = spokes==0?circles:0;
+    circle_orbit_diameter=hub_diameter/2+rim_radius;
+    circle_orbit_curcumference=pi*circle_orbit_diameter;
+
+    // Limit the circle size to 90% of the gear face.
+    circle_diameter=
+        min (
+            0.70*circle_orbit_curcumference/circles,
+            (rim_radius-hub_diameter/2)*0.9);
+
+    difference()
+    {
+        union ()
+        {
+            difference ()
+            {
+                //start with a plane toothed disk gear
+                linear_extrude_flat_option(flat=flat, height=rim_thickness, convexity=10, twist=twist)
+                gear_shape (
+                    number_of_teeth,
+                    pitch_radius = pitch_radius,
+                    root_radius = root_radius,
+                    base_radius = base_radius,
+                    outer_radius = outer_radius,
+                    half_thick_angle = half_thick_angle,
+                    involute_facets=involute_facets);
+
+                //if we have a spoked gear, then the gear body must be removed.
```

Comment:
Regarding this, I see that gears with a thinner body and reinforcing spokes apparently exist, so maybe we should have an option that allows for keeping the body in place even with spokes?

![image](https://user-images.githubusercontent.com/88355/126357839-0e726af0-b150-4f12-9e6e-0d301bde5335.png)

Reply:
neat, I guess we need spokes, circles and gear body at the same time. This will also need another feature I was anticipating called Z_SYMMETRY, which will center the gear rim, body, and hub on the XY plane. I will try to get this and the facet problem corrected in the next day. I thought that most of the rendering things like facets were handled by the global variables $fa, $fn. I just cranked mine up so I did not notice the poor rendering.

![wheel](https://user-images.githubusercontent.com/5097570/126375110-4514ebbf-cade-4b0d-bb91-58dac347fb58.jpg)

Reply:
I usually use the following values:
```scad
$fs = 0.4;
$fa = 1;
```

This usually yields pretty smooth surfaces, but when starting with a small cylinder diameter of `1`, it resulted in very few facets.

Reply:
Your trick with the cylinder worked like a charm. It now allows circles, spokes, and gear body, and also two flags for aligning the gear body and hub to the center of the rim height, which helps when you want the spokes centered on the gear body.

Reply:
An issue came up with the defaults. As it was, it required a gear_thickness, and then used that as a default value for the hub and the rim. I feel it should require a rim_thickness, and use that as the default for the gear and hub thickness. Reason being that not all gears will have a body (i.e. they have spokes instead), but all gears will have a rim, or it is not a gear. I suppose it could be written such that it used what ever you gave it (rim/hub/gear thickness) for all the unspecified ones.